File: linkedList/medium/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

from typing import Optional

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        new_head = head
        prev = None

        # reverse order
        while new_head:
            temp = new_head.next
            new_head.next = prev
            prev = new_head
            new_head = temp

        new_head = prev
        prev = None
        num = 1

        while new_head:
            temp = new_head.next
            # delete
            if num == n:
                new_head.next = None
            # reverse order
            else:
                new_head.next = prev
                prev = new_head
                
            new_head = temp
            num += 1

        return prev

File: linkedList/medium/19. Remove Nth Node From End of List.java
import linkedList.ListNode;

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // the distance between fast and slow is n+1
        // then slow should be in front of the deleting node
        ListNode dummy = new ListNode(0, head);
        ListNode fast = dummy, slow = dummy;
        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }

        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
        }

        // delete the target
        slow.next = slow.next.next;

        return dummy.next;
    }
}